awswrangler/opensearch/_write.py

```diff
@@ -0,0 +1,471 @@
+        which is used to execute the operation
+
+    Returns
+    -------
+    Dict[str, Any]
+        Response payload
+        https://opensearch.org/docs/opensearch/rest-api/document-apis/bulk/#response.
+
+    Examples
+    --------
+    Writing contents of CSV file
+
+    >>> import awswrangler as wr
+    >>> client = wr.opensearch.connect(host='DOMAIN-ENDPOINT')
+    >>> wr.opensearch.index_csv(
+    ...     client=client,
+    ...     path='docs.csv',
+    ...     index='sample-index1'
+    ... )
+
+    Writing contents of CSV file using pandas_kwargs
+
+    >>> import awswrangler as wr
+    >>> client = wr.opensearch.connect(host='DOMAIN-ENDPOINT')
+    >>> wr.opensearch.index_csv(
+    ...     client=client,
+    ...     path='docs.csv',
+    ...     index='sample-index1',
+    ...     pandas_kwargs={'sep': '|', 'na_values': ['null', 'none'], 'skip_blank_lines': True}
+    ... )
+    """
+    custom_pandas_params = {
+        'skip_blank_lines': True,
+        'na_filter': True # will generate Nan value for empty cells. We remove Nan keys in _df_doc_generator
+    }
+    pandas_kwargs.update(custom_pandas_params)
+    df = pd.read_csv(path, **pandas_kwargs)
+    return index_df(
+        client,
+        df=df,
+        index=index,
+        doc_type=doc_type,
+        **kwargs
+    )
+
+
+def index_df(
+    client: Elasticsearch,
+    df: pd.DataFrame,
+    index: str,
+    doc_type: Optional[str] = None,
+    **kwargs
+) -> Dict[str, Any]:
+    """Index all documents from a DataFrame to OpenSearch index.
+
+    Parameters
+    ----------
+    client : Elasticsearch
+        instance of elasticsearch.Elasticsearch to use.
+    df : pd.DataFrame
+        Pandas DataFrame https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html
+    index : str
+        Name of the index.
+    doc_type : str, optional
+        Name of the document type (only for Elasticsearch versions 5.x and older).
+    **kwargs :
+        KEYWORD arguments forwarded to :func:`~awswrangler.opensearch.index_documents`
+        which is used to execute the operation
+
+    Returns
+    -------
+    Dict[str, Any]
+        Response payload
+        https://opensearch.org/docs/opensearch/rest-api/document-apis/bulk/#response.
+
+    Examples
+    --------
+    Writing rows of DataFrame
+
+    >>> import awswrangler as wr
+    >>> import pandas as pd
+    >>> client = wr.opensearch.connect(host='DOMAIN-ENDPOINT')
+    >>> wr.opensearch.index_df(
+    ...     client=client,
+    ...     df=pd.DataFrame([{'_id': '1'}, {'_id': '2'}, {'_id': '3'}]),
+    ...     index='sample-index1'
+    ... )
+    """
+
+    return index_documents(
+        client=client,
+        documents=_df_doc_generator(df),
+        index=index,
+        doc_type=doc_type,
+        **kwargs
+    )
+
+
+def index_documents(
+    client: Elasticsearch,
+    documents: Union[Iterable[Dict[str, Any]], Iterable[Mapping[str, Any]]],
```

Comment:
Would only `Iterable[Mapping[str, Any]]` be enough? It should cover dicts as well

Reply:
changed to only `Iterable[Mapping[str, Any]]`
btw - I copied the original code from dynamodb module:
https://github.com/awslabs/aws-data-wrangler/blob/main/awswrangler/dynamodb/_write.py#L156